src/hyperion/experiment_plans/robot_load_then_centre_plan.py

```diff
@@ -203,45 +201,53 @@ def robot_load():
 
     yield from robot_load()
 
-    # XXX 1278 this effectively casts between unrelated types which doesn't have all
-    # attributes needed for downstream e.g. grid_width_microns
-    params_json = json.loads(parameters.json())
-    pin_centre_params = PinCentreThenXrayCentreInternalParameters(**params_json)
-
+    use_energy, use_resolution = yield from _get_updated_parameters_for_pin_and_xray(
+        params, composite
+    )
     yield from pin_centre_then_xray_centre_plan(
-        cast(GridDetectThenXRayCentreComposite, composite), pin_centre_params
+        cast(GridDetectThenXRayCentreComposite, composite),
+        params.pin_centre_then_xray_centre_params(
+            energy_ev=use_energy, ispyb_resolution=use_resolution
+        ),
     )
 
 
-def robot_load_then_centre(
-    composite: RobotLoadThenCentreComposite,
-    parameters: RobotLoadThenCentreInternalParameters,
-) -> MsgGenerator:
-    eiger: EigerDetector = composite.eiger
-
-    actual_energy_ev = 1000 * (
+def _get_updated_parameters_for_pin_and_xray(
+    params: RobotLoadThenCentre, composite: RobotLoadThenCentreComposite
+):
+    current_energy_ev = 1000 * (
         yield from read_energy(cast(SetEnergyComposite, composite))
     )
+    use_energy = params.demand_energy_ev or current_energy_ev
 
-    if not parameters.experiment_params.requested_energy_kev:
-        parameters.hyperion_params.detector_params.expected_energy_ev = actual_energy_ev
-    else:
-        parameters.hyperion_params.detector_params.expected_energy_ev = (
-            parameters.experiment_params.requested_energy_kev * 1000
-        )
-
-    wavelength_angstroms = convert_eV_to_angstrom(actual_energy_ev)
-    parameters.hyperion_params.ispyb_params.resolution = resolution(
-        parameters.hyperion_params.detector_params,
+    wavelength_angstroms = convert_eV_to_angstrom(use_energy)
+    use_resolution = resolution(
+        params.detector_params,
         wavelength_angstroms,
-        parameters.experiment_params.detector_distance,
+        params.detector_distance_mm
+        or (yield from bps.rd(composite.detector_motion.z.user_readback)),
```

Comment:
Nit: I think this internal yield adds a little too much complexity, I would suggest we just find the `det_distance` first then call `resolution`